### PR TITLE
Improved ask_for_experiment-Function

### PR DESCRIPTION
**Pull request by fdamken**
* Added SLURM ID to experiment ID from environment variables.
* Added more information to the experiments.
* Added possibility to show specific hyper-parameters.
* Implemented this to all calls of `ask_for_experiment(..)`.

### CODE REVIEW

File: Pyrado/pyrado/logger/experiment.py [hunk 1]
```diff
@@ -27,20 +27,21 @@
...
 import itertools
-import numpy as np
 import os
-import os.path as osp
-import torch as to
-import torch.nn as nn
-import yaml
 from copy import deepcopy
 from datetime import datetime
+from os import path as osp
```
Comment (famura) [inline Pyrado/pyrado/logger/experiment.py:33]: pls ceep the old import for osp

File: Pyrado/pyrado/utils/input_output.py [hunk 2]
```diff
@@ -27,15 +27,18 @@
...
-from typing import Sequence, Iterable, Callable, Any
+import typing
+from typing import Sequence, Iterable, Callable
 
+import numpy as np
 import pyrado
+from colorama import Fore, Style
 from pyrado.utils import run_once
 
+T = typing.TypeVar("T")
```
Comment (famura) [inline Pyrado/pyrado/utils/input_output.py:40]: is this necessary, can't We simply use Any?

File: Pyrado/scripts/evaluation/eval_policies_domain_grid_1d.py [hunk 3]
```diff
@@ -131,10 +143,10 @@
...
         ]
 
-    elif args.env_name == QQubeSwingUpSim.name:
-        env = QQubeSwingUpSim(dt=args.dt, max_steps=args.max_steps)
+    elif env_name == QQubeSwingUpSim.name:
+        env = QQubeSwingUpSim(dt=dt, max_steps=args.max_steps)
 
-        # param_spec['g'] = np.linspace(9.81*0.7, 9.81*1.3, num=11, endpoint=True)
+        param_spec["g"] = np.linspace(9.81 * 0.7, 9.81 * 1.3, num=11, endpoint=True)
         # param_spec['Rm'] = np.linspace(8.4*0.7, 8.4*1.3, num=11, endpoint=True)
```
Comment (famura) [inline Pyrado/scripts/evaluation/eval_policies_domain_grid_1d.py:150]: also run black on the rest (comment out for that)

File: Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Script to plot the results from the 2D domain parameter grid evaluations of a single policy.
+"""
+import os
+import os.path as osp
+
```
Comment (famura) [inline Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py:34]: import grouping (see above)

File: Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py:1]: please separate the addition of the new file from the actual content of this PR

**Review by famura — COMMENTED** [review]
Thank you for your contribution. Can you please say why you want to give the slrum ID a special emphasis?

**Comment by fdamken** [comment]
> Thank you for your contribution. Can you please say why you want to give the slrum ID a special emphasis?

It was useful to see which run finished just by looking at the IDs. But we can put it in the `extra_info`, of course.

File: Pyrado/pyrado/utils/input_output.py (hunk 2 above)
Comment (fdamken) [reply Pyrado/pyrado/utils/input_output.py:40]: Using the `TypeVar` has the advantage that it is known to e.g. PyCharm what the return type will be (in the case of asking for an experiment, PyCharm will know that `select_query` will return an `Experiment`).

File: Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py (hunk 5 above)
Comment (fdamken) [reply Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py:1]: whoopsy

File: Pyrado/pyrado/logger/experiment.py (hunk 1 above)
Comment (fdamken) [reply Pyrado/pyrado/logger/experiment.py]: Done with 275179c.

File: Pyrado/pyrado/logger/experiment.py [hunk 6]
```diff
@@ -50,7 +51,7 @@ class Experiment:
     This is a path-like object, and as such it can be used everywhere a normal path would be used.
 
     Experiment folder path:
-    <base_dir>/<env_name>/<algo_name>/<timestamp>--<extra_info>
+    <base_dir>/<env_name>/<algo_name>/<timestamp>--<extra_info>--<slurm_id>
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/logger/experiment.py]: Done with 275179c.

File: Pyrado/pyrado/logger/experiment.py [hunk 7]
```diff
@@ -243,6 +267,31 @@ def select_by_hint(exps: Sequence[Experiment], hint: str):
     return sl
 
 
+def create_experiment_formatter(
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/logger/experiment.py:263]: Done with 275179c.

File: Pyrado/pyrado/logger/experiment.py [hunk 8]
```diff
@@ -61,6 +62,7 @@ def __init__(
         exp_id: str = None,
         timestamp: datetime = None,
         base_dir: str = pyrado.TEMP_DIR,
+        include_slurm_id_if_possible: bool = True,
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/logger/experiment.py]: Done with 275179c.

File: Pyrado/pyrado/logger/experiment.py [hunk 9]
```diff
@@ -439,3 +493,18 @@ def load_dict_from_yaml(yaml_file: str) -> dict:
     with open(yaml_file, "r") as yaml_file:
         data = yaml.load(yaml_file, Loader=AugmentedSafeLoader)
     return data
+
+
+def load_hyperparameters(ex_dir, raise_error: bool = False) -> Union[dict, Optional[dict]]:
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/logger/experiment.py:498]: Done with 275179c.

File: Pyrado/pyrado/utils/argparser.py [hunk 10]
```diff
@@ -231,5 +231,11 @@ def get_argparser() -> argparse.ArgumentParser:
         action="store_false",
         help="the opposite of 'warmstart'",
     )
+    parser.add_argument(
+        "-p",
+        "--show_hyperparameters",
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/utils/argparser.py]: Done with 275179c and 8f53a75.

File: Pyrado/pyrado/utils/input_output.py [hunk 11]
```diff
@@ -27,15 +27,18 @@
 # POSSIBILITY OF SUCH DAMAGE.
 
 import contextlib
-import numpy as np
 import select
 import sys
-from colorama import Fore, Style
-from typing import Sequence, Iterable, Callable, Any
+import typing
+from typing import Sequence, Iterable, Callable
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/utils/input_output.py]: Done with 275179c – but shouldn't it be possible to include this into the Black formatter?

File: Pyrado/scripts/evaluation/eval_policies_domain_grid_1d.py (hunk 3 above)
Comment (fdamken) [reply Pyrado/scripts/evaluation/eval_policies_domain_grid_1d.py]: Done with 275179c.

File: Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py (hunk 5 above)
Comment (fdamken) [reply Pyrado/scripts/plotting/plot_policy_domain_grid_1d.py]: Done with 275179c.

File: Pyrado/pyrado/utils/argparser.py [hunk 12]
```diff
@@ -231,5 +231,11 @@ def get_argparser() -> argparse.ArgumentParser:
...
+        action="append",
+        help="The hyperparameters to show in the ask-for-experiment dialog. Use this parameter multiple times to show multiple hyperparameters.",
```
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/utils/argparser.py]: Done with 7b85a6b.

File: Pyrado/pyrado/logger/experiment.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fdamken) [reply Pyrado/pyrado/logger/experiment.py]: I put it in the `extra_info` field now.

File: Pyrado/pyrado/utils/input_output.py (hunk 2 above)
Comment (famura) [reply Pyrado/pyrado/utils/input_output.py:40]: I see, then we keep it

File: Pyrado/scripts/evaluation/eval_lfi_posterior_rollouts.py [hunk 13]
```diff
@@ -61,7 +61,7 @@
     num_ml_samples = args.num_rollouts_per_config
 
     # Get the experiment's directory to load from
-    ex_dir = ask_for_experiment() if args.dir is None else args.dir
+    ex_dir = ask_for_experiment(show_hyper_parameters=args.show_hparams) if args.dir is None else args.dir
```
Comment (famura) [inline Pyrado/scripts/evaluation/eval_lfi_posterior_rollouts.py:64]: also rename this show_hyper_parameters to show_hparams

File: Pyrado/pyrado/logger/experiment.py [hunk 14]
```diff
@@ -71,8 +73,25 @@ def __init__(
         :param exp_id: combined timestamp and extra_info, usually the final folder name
         :param timestamp: experiment creation timestamp
         :param base_dir: base storage directory
+        :param include_slurm_id: If a SLURM ID is present in the environment variables, include them in the experiment ID.
```
Comment (famura) [inline Pyrado/pyrado/logger/experiment.py:76]: IMO, the advantage of tracking the slurm ID is not as important as the loss in generatlity. Thus, I would like to ask you to remove the include_slurm_id and everything associated to it

File: Pyrado/pyrado/logger/experiment.py [hunk 15]
```diff
@@ -132,10 +143,16 @@ def matches(self, hint: str) -> bool:
             return self.env_name == env_name and self.algo_name == algo_name and self.exp_id == eid
 
 
-def setup_experiment(env_name: str, algo_name: str, extra_info: str = None, base_dir: str = pyrado.TEMP_DIR):
+def setup_experiment(
+    env_name: str,
+    algo_name: str,
+    extra_info: str = None,
+    base_dir: str = pyrado.TEMP_DIR,
+    include_slurm_id_if_possible: bool = True,
```
Comment (famura) [inline Pyrado/pyrado/logger/experiment.py:151]: also remove here

File: Pyrado/pyrado/logger/experiment.py [hunk 16]
```diff
@@ -243,6 +260,38 @@ def select_by_hint(exps: Sequence[Experiment], hint: str):
     return sl
 
 
+def create_experiment_formatter(
+    show_hparams: Optional[List[str]] = None, show_extra_info: bool = True
+) -> Callable[[Experiment], str]:
+    """
+    Returns an experiment formatter (i.e. a function that takes an experiment and produces a string) to be used in the
```
Comment (famura) [inline Pyrado/pyrado/logger/experiment.py:267]: good, thanks

File: Pyrado/pyrado/logger/experiment.py [hunk 17]
```diff
@@ -439,3 +493,25 @@ def load_dict_from_yaml(yaml_file: str) -> dict:
     with open(yaml_file, "r") as yaml_file:
         data = yaml.load(yaml_file, Loader=AugmentedSafeLoader)
     return data
+
+
+def load_hyperparameters(ex_dir, raise_error: bool = False) -> Union[dict, Optional[dict]]:
+    """
+    Loads the hyperparameters-dict from the given experiment directory. The hyperparameters file is assumed to be
+    named :code:`hyperparams.yaml`.
```
Comment (famura) [inline Pyrado/pyrado/logger/experiment.py:501]: good, thanks

**Review by famura — CHANGES_REQUESTED** [review]
good changes, decided against axplicit tarcking of the slurm ID. I'd like to merge the other changes though

File: Pyrado/pyrado/logger/experiment.py [hunk 18]
```diff
@@ -132,10 +143,16 @@ def matches(self, hint: str) -> bool:
...
+    env_name: str,
+    algo_name: str,
+    extra_info: str = None,
+    base_dir: str = pyrado.TEMP_DIR,
+    include_slurm_id_if_possible: bool = True,
+):
     """ Setup a new experiment for recording. """
     # Create experiment object
-    exp = Experiment(env_name, algo_name, extra_info, base_dir=base_dir)
+    exp = Experiment(env_name, algo_name, extra_info, base_dir=base_dir, include_slurm_id=include_slurm_id_if_possible)
```
Comment (famura) [inline Pyrado/pyrado/logger/experiment.py:155]: remove `if_possible`

**Comment by fdamken** [comment]
Moved to another PR for reduced diff.
